### PR TITLE
[DTensor] Add sharding strategies for pooling ops

### PR DESCRIPTION
**Pull request by pianpwk**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #177471
* #177376
* #176992
* #176991
* #176955
* __->__ #176821

The validation results show missing rules, but pretty sure they're false positives.
High-level logic:
- 2 types of pooling, avg & max. avg propagates P(avg), P(sum). max does not propagate P(max), because it returns (values, indices) and indices need replication.
- pooling is 1d, 2d, or 3d ("spatial rank").
- Shard propagates on batched dimensions; dim 0 is always batched; dim 1 is batched when input.ndim >= spatial rank + 2 (pool1d on 3d, pool2d on 4d, pool3d on 5d)

Validation results (limited maxpool to 30 samples because took too long):
```
(pytorch-3048) [pianpwk@devvm3048.dkl0 /data/users/pianpwk/ptclone/pytorch (59930bf2)]$ python -m torch.distributed.tensor._ops.strategy_validation --op "nn.functional.adaptive_avg_pool2d,nn.functional.adaptive_avg_pool3d,nn.functional.avg_pool2d,nn.functional.avg_pool3d"
Testing ops: nn.functional.adaptive_avg_pool2d, nn.functional.adaptive_avg_pool3d, nn.functional.avg_pool2d, nn.functional.avg_pool3d
Device: cuda, Dtype: torch.float32, World size: 2

[1/4] nn.functional.adaptive_avg_pool2d — Samples: 10, Combinations: 640
----------------------------------------------------------------------

Possibly missing (valid in ground truth but no DTensor rule)

  [aten._adaptive_avg_pool2d.default]
    P(max) -> P(max)
    P(min) -> P(min)
    S(1) -> S(1)
    S(2) -> S(2)
    S(3) -> S(3)

[2/4] nn.functional.adaptive_avg_pool3d — Samples: 11 (1 skipped), Combinations: 882
----------------------------------------------------------------------
  aten._adaptive_avg_pool3d.default: 29 correct, 0 incorrect, 6 missing
  aten.mean.dim: 4 correct, 0 incorrect, 2 missing

Possibly missing (valid in ground truth but no DTensor rule)

  [aten._adaptive_avg_pool3d.default]
    P(max) -> P(max)
    P(min) -> P(min)
    S(1) -> S(1)
    S(2) -> S(2)
    S(3) -> S(3)
    S(4) -> S(4)

  [aten.mean.dim]
    P(avg) -> P(avg)
    P(sum) -> P(sum)

[3/4] nn.functional.avg_pool2d — Samples: 7, Combinations: 504
----------------------------------------------------------------------

Possibly missing (valid in ground truth but no DTensor rule)

  [aten.avg_pool2d.default]
    S(2) -> S(2)
    S(3) -> S(3)

[4/4] nn.functional.avg_pool3d — Samples: 7 (1 skipped), Combinations: 630
----------------------------------------------------------------------

Possibly missing (valid in ground truth but no DTensor rule)

  [aten.avg_pool3d.default]
    S(2) -> S(2)
    S(3) -> S(3)
    S(4) -> S(4)

======================================================================
Summary
======================================================================
Op                                 Correct  Incorrect  Missing    Time
----------------------------------------------------------------------
nn.functional.adaptive_avg_pool2d       31          0        5    10.1s
nn.functional.adaptive_avg_pool3d       33          0        8    13.7s
nn.functional.avg_pool2d                21          0        2     7.6s
nn.functional.avg_pool3d                19          0        3     9.1s
----------------------------------------------------------------------
Total                                  104          0       18    40.6s

pytorch-3048) [pianpwk@devvm3048.dkl0 /data/users/pianpwk/ptclone/pytorch (59930bf2)]$ python -m torch.distributed.tensor._ops.strategy_validation --op "nn.functional.adaptive_max_pool2d,nn.functional.adaptive_max_pool3d,nn.functional.fractional_max_pool2d,nn.functional.fractional_max_pool3d,nn.functional.max_pool2d,nn.functional.max_pool3d" --max-samples 30
Testing ops: nn.functional.adaptive_max_pool2d, nn.functional.adaptive_max_pool3d, nn.functional.fractional_max_pool2d, nn.functional.fractional_max_pool3d, nn.functional.max_pool2d, nn.functional.max_pool3d
Device: cuda, Dtype: torch.float32, World size: 2

[1/6] nn.functional.adaptive_max_pool2d — Samples: 24, Combinations: 1536
----------------------------------------------------------------------

Possibly missing (valid in ground truth but no DTensor rule)

  [aten.adaptive_max_pool2d.default]
    P(avg) -> (P(avg), P(avg))
    P(avg) -> P(avg)
    P(max) -> (P(max), P(max))
    P(max) -> P(max)
    P(min) -> (P(min), P(min))
    P(min) -> P(min)
    P(sum) -> P(sum)
    S(1) -> P(max)
    S(1) -> S(1)
    S(2) -> P(max)
    S(2) -> S(2)
    S(3) -> P(max)

[2/6] nn.functional.adaptive_max_pool3d — Samples: 20, Combinations: 1620
----------------------------------------------------------------------

Possibly missing (valid in ground truth but no DTensor rule)

  [aten.adaptive_max_pool3d.default]
    P(avg) -> (P(avg), P(avg))
    P(avg) -> P(avg)
    P(max) -> (P(max), P(max))
    P(max) -> P(max)
    P(min) -> (P(min), P(min))
    P(min) -> P(min)
    P(sum) -> P(sum)
    S(1) -> P(max)
    S(1) -> S(1)
    S(2) -> P(max)
    S(2) -> S(2)
    S(3) -> P(max)
    S(3) -> S(3)
    S(4) -> P(max)
    S(4) -> S(4)

[3/6] nn.functional.fractional_max_pool2d — skipped: no DTensor support

[4/6] nn.functional.fractional_max_pool3d — skipped: no DTensor support

[5/6] nn.functional.max_pool2d — Samples: 30, Combinations: 2160
----------------------------------------------------------------------

Possibly missing (valid in ground truth but no DTensor rule)

  [aten.max_pool2d_with_indices.default]
    P(max) -> P(max)
    S(2) -> P(max)
    S(2) -> S(2)
    S(3) -> S(3)

[6/6] nn.functional.max_pool3d — Samples: 30, Combinations: 2700
----------------------------------------------------------------------

Possibly missing (valid in ground truth but no DTensor rule)

  [aten.max_pool3d_with_indices.default]
    P(max) -> P(max)
    S(2) -> P(max)
    S(2) -> (R, R)
    S(2) -> R
    S(2) -> S(2)
    S(3) -> S(3)
    S(4) -> S(4)

======================================================================
Summary
======================================================================
Op                                 Correct  Incorrect  Missing    Time
----------------------------------------------------------------------
nn.functional.adaptive_max_pool2d       28          0       12    25.0s
nn.functional.adaptive_max_pool3d       24          0       15    27.0s
nn.functional.max_pool2d                30          0        4    34.7s
nn.functional.max_pool3d                30          0        7    43.7s
----------------------------------------------------------------------
Total                                  112          0       38   130.3s
```

### CODE REVIEW

**Comment by wconstab** [comment]
were there a lot of false positives on the normal run (saw you used '--incorrect_only' this time)
- nit: would be nice to remove the 'missing' column from the summary table when doing --incorrect_only, as a careless reader might look at that and get false confidence

File: torch/distributed/tensor/_ops/_math_ops.py [hunk 1]
```diff
@@ -775,6 +775,76 @@ def linalg_replicate_strategy(op_schema: OpSchema) -> OpStrategy:
...
+    aten.fractional_max_pool3d.default,
+    aten.max_pool1d_with_indices.default,
+    aten.max_pool2d_with_indices.default,
+    aten.max_pool3d_with_indices.default,
+]
+
+
+@register_op_strategy(
+    AVG_POOL_OPS + MAX_POOL_OPS,
+    schema_info=RuntimeSchemaInfo(1),
```
Comment (wconstab) [inline torch/distributed/tensor/_ops/_math_ops.py:821]: i'd appreciate a comment anytime we register RuntimeSchemaInfo saying something about why

File: torch/distributed/tensor/_ops/_math_ops.py [hunk 2]
```diff
@@ -775,6 +775,76 @@ def linalg_replicate_strategy(op_schema: OpSchema) -> OpStrategy:
...
+)
+def pooling_strategy(op_schema: OpSchema) -> OpStrategy:
+    input_strategy = cast(OpStrategy, op_schema.args_schema[0])
+    mesh = input_strategy.mesh
+    num_outputs = 2 if op_schema.op in MAX_POOL_OPS else 1
+    num_inputs = len(op_schema.args_strategy) + len(op_schema.kwargs_strategy)
```
Comment (wconstab) [inline torch/distributed/tensor/_ops/_math_ops.py:827]: what are the cases where num_inputs > 1? (at least > 1 tensor, i mean)

**Review by wconstab — APPROVED** [review]
